src/util.js

```diff
@@ -0,0 +1,44 @@
+import axios from 'axios'
+
+export function getCodeVerifier() {
+  const lowerCase = 'abcdefghijklmnopqrstuvwxyz'
+  const upperCase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+  const numbers = '1234567890'
+  const specialCharacters = "-._~"
+  const characters = lowerCase + upperCase + numbers + specialCharacters
+
+  const minLength = 43
+  const maxLength = 128
+  const stringLength = Math.floor(Math.random() * (maxLength - minLength + 1)) + minLength
+
+  let string = ''
+  for (var i = 0; i < stringLength; i++) {
+    const randomIndex = Math.floor(Math.random() * characters.length)
+    string += characters.charAt(randomIndex)
+  }
+
+  return string;
+}
+
+export function getAuthUrl() {
+    const codeVerifier = window.localStorage.getItem('DW-CODE-VERIFIER')
```

Comment:
Move `DW-CODE-VERIFIER` into a `const` since you are using it in multiple places

Reply:
It has to be stored in local storage because auth redirects would delete any stored variables.

There's now a `getCodeVerifier` function in `util` that reads and returns the value stored in local storage thereby reducing code repetition.